### PR TITLE
always resolve to universal regions if possible

### PR DESCRIPTION
**Pull request by aliemjay**
`RegionConstraintCollector::opportunistic_resolve_var`, which is used in canonicalization and projection logic, doesn't resolve the region var to an equal universal region. So if we have equated `'static == '1 == '2`, it doesn't resolve `'1` or `'2` to `'static`. Now it does!

Addresses review comment https://github.com/rust-lang/rust/pull/107376#discussion_r1093233687.

r? @lcnr

### CODE REVIEW

**Comment by lcnr** [comment]
that merge conflict is unfortunate :sweat_smile: 

r=me after perf

**Comment by aliemjay** [comment]
The original change introduced a new unsoundness (see last commits). It is caused by resolving a region variable to a placeholder region from a universe that it can't name. I originally thought this is not a problem because region resolution doesn't rely on `unification_table` and we would eventually have a universe error, but it tuned out that canonicalization uses opportunistic resolution on query region constraints opening a loophole in this logic.

Theoretically, it is a preexisting bug because in the previous setup we could resolve a variable to another variable of higher universe, and thus allowing it to name more regions, but I can't think of an example where we unify two region variables from different universes inside a canonical query in order to exploit this.

**Comment by lcnr** [comment]
@bors try @rust-timer queue

I am unhappy with bfd35016e41724a370828dc400e6b12e1aaf9201 as it feels fairly brittle. Can look into this myself at some point, but how difficult is it to 1. eagerly error when unifying a region with a placeholder from a higher universe and 2. when unifying 2 regions `'a` and `'b`, set their universes to `min(uv('a), uv('b))`.

**Comment by aliemjay** [comment]
> I am unhappy with [bfd3501](https://github.com/rust-lang/rust/commit/bfd35016e41724a370828dc400e6b12e1aaf9201) as it feels fairly brittle. Can look into this myself at some point, but how difficult is it to 1. eagerly error when unifying a region with a placeholder from a higher universe and 2. when unifying 2 regions `'a` and `'b`, set their universes to `min(uv('a), uv('b))`.

I don't see a problem with the current approach. `unification_table` stores what vars/values we unified and only when we use it for anything useful we should concern ourselves with universes. I think the alternative approach would be unnecessarily painful and probably a perf hit.

Anway, I tried to implement alternative approach in https://github.com/aliemjay/rust/tree/resolve-var-region-alt but I don't really like it especially how the universes of region variables change during inference.

**Comment by aliemjay** [comment]
cc https://github.com/rust-lang/rust/pull/108867

File: compiler/rustc_infer/src/infer/region_constraints/mod.rs [hunk 1]
```diff
@@ -633,28 +633,22 @@ impl<'tcx> RegionConstraintCollector<'_, 'tcx> {
...
-                let unified_region = self.unification_table().probe_value(rid);
-                unified_region.0.unwrap_or_else(|| {
-                    let root = self.unification_table().find(rid).vid;
-                    tcx.mk_re_var(root)
-                })
-            }
-            _ => region,
+        let mut ut = self.unification_table_mut(); // FIXME(rust-lang/ena#42): unnecessary mut
+        let root_vid = ut.find(vid).vid;
+        let resolved = ut.probe_value(root_vid).0.unwrap_or_else(|| tcx.mk_re_var(root_vid));
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/region_constraints/mod.rs:645]: this is the only use of `probe_value` for regions :thinking: so yeah I agree that this is fine.

Can you add a comment to `UnifiedRegion` that the assigned value may not be nameable by the region var?

Or even better, change unified region to 
```rust
pub(super?) struct UnifiedRegion {
    resolved: Option<Region<'tcx>>,
}

impl UnifiedRegion {
    /// Something something the resolved region may not be nameable by
    /// this region variable.
    pub fn get_resolved_region_ignoring_universes(self) -> Option<Region<'tcx>> {
        self.resolved
    }
}
```

**Review by lcnr — APPROVED** [review]
looked a bit more into this and also at #108867

I think I agree with you now that this approach is nicer but thought of a way to make this harder to misuse by accident in the future.

r=me after this nit

**Comment by aliemjay** [comment]
@lcnr one more commit to review :sweat_smile:

**Comment by aliemjay** [comment]
I'll pull it to another PR. It's not that important.

**Comment by nnethercote** [comment]
The improvements greatly outweight the miniscule regressions here. The cycles and wall-time numbers for `wg-grammar` are especially nice.

@rustbot label: +perf-regression-triaged

**Pull request merged**